aten/src/ATen/native/Sorting.cpp

```diff
@@ -183,148 +293,51 @@ std::tuple<Tensor, Tensor> kthvalue(
-  auto quantiles = values_below.lerp_(values_above, weights);
-
-  // when the q tensor is not a scalar, numpy will prepend a new dimension
-  // to contain the quantiles. This code ensures we follow the same order
-  if (q.dim() > 0) {
-    quantiles.unsqueeze_(0);
-    ++dim;
-    std::vector<int64_t> numpy_dim_order;
-    for (int64_t i = 0; i < quantiles.dim(); ++i) {
-      numpy_dim_order.push_back(i);
-    }
-    std::iter_swap(numpy_dim_order.begin(), numpy_dim_order.begin() + dim);
-    quantiles = quantiles.permute(numpy_dim_order);
-  }
-
-  return quantiles;
-}
-
 Tensor& quantile_out(
     Tensor& out,
     const Tensor& self,
     const Tensor& q,
-    c10::optional<int64_t> _dim,
+    optional<int64_t> _dim,
     bool keepdim) {
-  auto out_shape = quantile_out_shape(self, q, _dim, keepdim);
-
-  TORCH_CHECK(out.sizes().vec() == out_shape,
-      "expected out shape to be ", out_shape, " but got ", out.sizes().vec());
-  TORCH_CHECK(self.scalar_type() == out.scalar_type(), 
-      "out tensor must be same dtype as the input tensor");
-  TORCH_CHECK(self.device() == out.device(), 
-      "out tensor must be on the same device as the input tensor");
-
-  if (q.numel() == 0) {
-    return out;
-  }
-
-  out.copy_(quantile_impl(self, q, _dim, keepdim).reshape(out_shape));
-
+  quantile_impl(out, self, q, std::move(_dim), keepdim);
   return out;
 }
 
 Tensor& quantile_out(
     Tensor& out,
     const Tensor& self,
     double q,
-    c10::optional<int64_t> _dim,
+    optional<int64_t> _dim,
     bool keepdim) {
-  TORCH_CHECK(q >= 0 && q <= 1, "q must be in the range [0, 1]");
-  return at::quantile_out(out, self, at::scalar_tensor(q, self.options()), _dim, keepdim);
+  TORCH_CHECK(
+      q >= 0 && q <= 1, "quantile() q must be in the range [0, 1] but got ", q);
+  return at::quantile_out(
+      out,
+      self,
+      at::scalar_tensor(q, self.options()),
+      std::move(_dim),
+      keepdim);
 }
 
 Tensor quantile(
     const Tensor& self,
     const Tensor& q,
-    c10::optional<int64_t> _dim,
+    optional<int64_t> _dim,
     bool keepdim) {
-  auto out_shape = quantile_out_shape(self, q, _dim, keepdim);
-
-  if (q.numel() == 0) {
-    return at::empty(out_shape, self.options());
-  }
-
-  return quantile_impl(self, q, _dim, keepdim).reshape(out_shape);
+  Tensor out = at::empty({0}, self.options());
+  quantile_impl(out, self, q, std::move(_dim), keepdim);
+  return out;
 }
 
 Tensor quantile(
     const Tensor& self,
     double q,
-    c10::optional<int64_t> _dim,
+    optional<int64_t> _dim,
     bool keepdim) {
-  TORCH_CHECK(q >= 0 && q <= 1, "q must be in the range [0, 1]");
-  return at::quantile(self, at::scalar_tensor(q, self.options()), _dim, keepdim);
+  TORCH_CHECK(
+      q >= 0 && q <= 1, "quantile() q must be in the range [0, 1] but got ", q);
+  return at::quantile(
+      self, at::scalar_tensor(q, self.options()), std::move(_dim), keepdim);
```

Comment:
Since we are creating a scalar_tensor to wrap q anyway, should we accept a list as well? As in float[1] q in native_functions?

Reply:
Sure.